README.md
# sensor-block
Auto-detects connected i2c sensors and publsihes data on HTTP or MQTT.

## Features
- Uses Indusrial IO (iio) to communicate with sensors, utilizing drivers already in the kernel to talk to the sensor directly
- Data published via mqtt and/or http
- Provides raw sensor data or "tranforms" the sensor data into a more standardized format 
- json output can either be one measurement per sensor or all sensor fields in one list 

## Usage

**Docker compose file**

To use this image, create a container in your `docker-compose.yml` file as shown below:

```
services:
  sensor:
    image: bh.cr/balenalabs/sensor-<arch>
    privileged: true
    labels:
      io.balena.features.kernel-modules: '1'
      io.balena.features.sysfs: '1'
      io.balena.features.supervisor-api: '1'
    expose:
      - '7575'  # Only needed if using  http server
```

## Overview/Compatibility
This block utilizes the [Linux Industrial I/O Subsystem](https://wiki.analog.com/software/linux/docs/iio/iio) ("iio") which is a kernel subsystem that allows for ease of implementing drivers for sensors and other similar devices such as ADCs, DACs, etc.  You can see a list of available iio drivers [here](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/drivers/iio?h=linux-5.4.y) but in order to save space, most OSes do not include all these drivers. The easiest way to check if your board supports a driver is to use the modinfo command on a running device. For instance:
```
modinfo ti-ads1015
```
This command searches the running kernel for all the drivers it includes and prints out a description of any that are found to match. (Use the info in the Kconfig file in each folder of the driver list above to find the proper driver name to use with this command) BalenaOS for Raspberry Pi 3 includes a small subset of these drivers which are listed in the table below. As we continue to test sensors and improve the block, more sensors should be supported and the chart will be updated accordingly. Also note that as kernel and OS versions change, the supported drivers may change somewhat as well. It's best therefore to use modinfo as described above to test compatibility with any sensor being considered for use with this block.

| Sensor Model | Sensor Name | Driver Name | Address(es) | Tested? |
| ------------ | ----------- | ----------- | ----------- | ------- |
| AD5301 | Analog Devices AD5446 and similar single channel DACs driver, TI DACs | ads5446 | 0xC, 0xD, 0xE, 0xF | Not tested |
| APDS9960 | Avago APDS9960 gesture/RGB/ALS/proximity sensor | apds9960 | 0x39 | Yes, NOT working |
| BME680 | Bosch Sensortec BME680 sensor | bme680 | 0x76, 0x77 | Yes, works |
| BMP180 | Bosch Sensortec BMP180 sensor | bmp280 | 0x77 | Not tested |
| BMP280 | Bosch Sensortec BMP280 sensor | bmp280 | 0x76, 0x77 | Yes, works |
| BME280 | Bosch Sensortec BME280 sensor | bmp280 | 0x76, 0x77 | Yes, works |
| HDC1000 | TI HDC100x relative humidity and temperature sensor | hdc100x | 0x40 - 0x43 | Not tested |
| HTU21 | Measurement Specialties HTU21 humidity & temperature sensor | htu21 | 0x40 | Yes, works |
| MS8607 | TE Connectivity PHT sensor | htu21 | 0x40, 0x76 | Yes, works partially (no pressure reading) |
| MCP342x | Microchip Technology MCP3421/2/3/4/5/6/7/8 ADC | mcp3422 | 0x68 - 0x6F | Not tested |
| ADS1015 | Texas Instruments ADS1015 ADC | ti-ads1015 | 0x48 - 0x4B | Yes, NOT working |
| TSL4531 | TAOS TSL4531 ambient light sensors | tsl4531 | 0x29 | Not tested |
| VEML6070 | VEML6070 UV A light sensor | veml6070 | 0x38, 0x39 | Yes, works |

By default, the block searches for sensors on SMBus number 1 (/dev/i2c-1) however you can set the bus number (an integer value) using the `BUS_NUMBER` service variable.

### Publishing Data

The sensor data is available in json format either as an mqtt payload and/or via the built-in http server. If you include an mqtt broker container named "mqtt" in your application, the block will automatically publish to that. If you provide an address for the `MQTT_ADDRESS` service variable, it will publish to that broker instead. The default interval for publishing data is eight seconds, which you can override with the `MQTT_PUB_INTERVAL` service variable by providing a value in seconds. The default topic is set to `sensors` (which is compatible with the [connector block](https://github.com/balena-labs-projects/connector#mqtt) ) which can be overridden by setting the `MQTT_PUB_TOPIC` service variable.

If no mqtt broker is set, the http server will be available on port 7575. To force the http server to be active even with mqtt, set the `ALWAYS_USE_HTTPSERVER` service variable to True.

The http data defaults to only be available to other containers in the application via `sensor:7575` - if you want this to be available externally, you'll need to map port 7575 to an external port in your docker-compose file.

## Data

The JSON for raw sensor data is available in one of two formats and is determined by the `COLLAPSE_FIELDS` service variable. The default value of `0` (zero) causes each sensor to output a separate measurement:
```
[{"measurement": "htu21", "fields": {"humidityrelative": 29700, "temp": 23356}}, {"measurement": "bmp280", "fields": {"pressure": 99.911941406, "temp": 23710}}, {"measurement": "short_UUID", "fields": {"short_uuid": "0f33e61"}}]
```

Changing `COLLAPSE_FIELDS` to `1` collapses all of the field values into one list like this:
```
{"short_uuid": "0f33e61", "humidityrelative": 29700, "temp": 23356, "pressure": 99.911941406}
```
Note that if two sensors output the same field name, it will only show up once in the list from one of the sensors. This feature is best used when the field names from sensors do not overlap.

The above examples display the raw data from the sensor as exposed by the driver, which is the default setting for the block. In many cases, the values and names need transformations to be useful. You can change the `RAW_VALUES` service variable from the default value of `1` to `0` (zero) to output transformed data instead. All the transformations are defined per-sensor in the `transformations.py` file which you can edit to your needs. We've included some basic ones for you. For example, here is the raw output of a bme680:
```
{"short_uuid": "0f33e61", "humidityrelative": 35.524, "pressure": 1005.48, "resistance": 52046.0, "temp": 24620.0}
```
Here is the transformed output with `RAW_VALUES` set to `0`:
```
{"short_uuid": "0f33e61", "pressure": 1005.4, "resistance": 65454.0, "humidity": 35.541, "temperature": 24.54}
```

When using transformed data outputs, you can change the temperature field from Celsius to Farenheit by setting the `TEMP_UNIT` variable to `F` (the default is `C`)

Note that the device's short UUID is always included in the data output, which can be useful for aggregating data from multiple devices. The short UUID is a string value so it will not show up in Grafana dashboards based on the [dashboard block](https://github.com/balena-labs-projects/dashboard).

## Use with other blocks

The sensor block works well with our [connector block](https://github.com/balena-labs-projects/connector) and [dashboard block](https://github.com/balena-labs-projects/dashboard). See the latest version of [balenaSense](https://github.com/balena-labs-projects/balena-sense) for an example of using all of these blocks together to read data from one or more sensor.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v0.0.4
## (2023-07-21)

* Delete docker-compose.yml [Alan Boris]

# v0.0.3
## (2023-07-21)

* Fix typo, force rebuild [Alan Boris]

# v0.0.2
## (2022-11-16)

* moving from balenablocks to balena-labs-projects [Flynn Joffray]
* Pizero build script [Phil Wilson]
* Add repo file [Phil Wilson]
* Added build script [Phil Wilson]


idetect.py
#!/usr/bin/env python3
import subprocess
import sys
from smbus2 import SMBus
import errno
import os
import iio
import json
import time

# Decimal address values
devices = {
12:     "ad5446",
13:     "ad5446",
14:     "ad5446",
15:     "ad5446",
41:     "tsl4531",
56:     "veml6070",
57:     "multiple",
64:     "multiple",
65:     "hdc100x",
66:     "hdc100x",
67:     "hdc100x",
64:     "multiple",
72:     "ti-ads1015",
73:     "ti-ads1015",
74:     "ti-ads1015",
75:     "ti-ads1015",
104:    "mcp3422",
105:    "mcp3422",
106:    "mcp3422",
107:    "mcp3422",
108:    "mcp3422",
109:    "mcp3422",
110:    "mcp3422",
111:    "mcp3422",
118:    "multiple",
119:    "multiple"
}

del_drivers = {
"ad5446.c": "ad5446.c",
"apds9960": "apds9960",
"bme680": "bme680-i2c",
"bmp280": "bmp280-i2c",
"dht11": "dht11",
"hdc100": "hdc100",
"htu21": "htu21",
"mcp320x": "mcp320x",
"mcp3422": "mcp3422",
"ti-ads1015": "ti-ads1015",
"tsl4531": "tsl4531",
"veml6070": "veml6070"
}

bosch_chip_id = {
0x61: "bme680",
0x55: "bmp180",
0x58: "bmp280",
0x60: "bme280"
}

def read_chip_id(bus, device, loc):
    chip_id = -1
    try:
        chip_id = bus.read_byte_data(device, loc)
    except Exception as e:
            print("Error while reading chip ID of device at address {0}: {1}".format(e, hex(device)))
    
    if chip_id == -1:  # Wait a tiny bit and try one more time
        time.sleep(2)
        try:
            chip_id = bus.read_byte_data(device, loc)
        except Exception as e:
            print("Error again while reading chip ID of device at address {0}: {1}".format(e, hex(device)))

    return chip_id

def detect_iio_sensors():
    bus_number = int(os.getenv('BUS_NUMBER', '1'))  # default 1 indicates /dev/i2c-1
    bus = SMBus(bus_number)
    device_count = 0
    active = []
    chip_id = 0

    print("======== Searching i2c bus for devices... ========")
    for device in range(3, 128):
        try:
            bus.write_byte(device, 0)
            print("Found device at {0}".format(hex(device)))
            active.append(device)
            device_count = device_count + 1
        except IOError as e:
            if e.errno != errno.EREMOTEIO:
                #print("Warning: {0} on address {1}".format(e, hex(device)))
                if e.errno == 16:  # device busy
                    print("Found (busy) device at {0}".format(hex(device)))
                    active.append(device)
                    device_count = device_count + 1
        except Exception as e: # exception if read_byte fails
            print("Error while searching for devices on i2c: {0} on address {1}".format(e, hex(device)))

    # use i2cdetect to see if HTU21D is on address 64 (0x40) since nothing else seems to detect it
    if 64 not in active:
        p = subprocess.Popen(['i2cdetect', '-y','1'],stdout=subprocess.PIPE,)
        for i in range(0,6):
            line = str(p.stdout.readline())
            #print(line)
            if line.find("40: 40") > 0:
                active.append(64)
                print("Found device (via i2cdetect) at 0x40")
    if device_count > 0:
        # We want to remove any existing devices if they are present
        # First uninstantiate the i2c bus
        print("======== Removing existing devices from the i2c bus... ========")
        for device in active:
            print("Deleting device found at {0}.".format(hex(device)))
            delete_device = "echo {0} > /sys/bus/i2c/devices/i2c-1/delete_device".format(hex(device))
            os_out = os.system(delete_device)
            if os_out > 0:
                print("Delete device {0} exit code: {1}".format(hex(device), os_out))

        print("======== Unloading any existing modules... ========")
        # Next unload any present devices using modprobe -rv
        output = subprocess.check_output("lsmod").decode()  # TODO: replace check_output with run variant
        d = []
        i = 0
        for line in output.split('\n'):
            i = i + 1
            if i > 1:
                line = line.strip()
                if line:
                    lsmod_module = line.split()   # splits string at whitespaces
                    find_underscore = lsmod_module[0].find("_")   # equals -1 if no underscore
                    if find_underscore > 0:
                        mod_name = lsmod_module[0][0:find_underscore]  # strip underscore and everything following
                    else:
                        mod_name = lsmod_module[0]
                    d.append(mod_name)
        # remove duplicates from list
        dd = []
        [dd.append(x) for x in d if x not in dd]
        # find in dict
        #print("Currently loaded modules: {0}".format(dd))
        for x in dd:
            if x in del_drivers:
                print("Unloading module {0} as {1}.".format(x, del_drivers[x]))
                subprocess.run(["modprobe", "-r", x])

        # Remove unrecognized devices
        #print("Active: {0}".format(active))
        #print("Keys: {0}".format(devices.keys()))
        new_active = []
        for x in active:
            if x in devices.keys():
                new_active.append(x)
            else:
                print("Device at {0} not in known supported drivers.".format(hex(x)))

        print("New active: {0}".format(new_active))

        # Now, load all the devices found
        print("======== Loading devices found... ========")
        subprocess.run(["modprobe", "crc8"])
        subprocess.run(["modprobe", "industrialio"])
        new_active_count = 0
        for device in new_active:
            if devices[device] != "multiple":
                print("Loading device {0} on address {1}.".format(devices[device], hex(device)))
                subprocess.run(["modprobe", devices[device]])
                new_device = "echo {0} {1} > /sys/bus/i2c/devices/i2c-1/new_device".format(devices[device], hex(device))
                os_out = os.system(new_device)
                if os_out > 0:
                    print("New device {0} exit code: {1}".format(hex(device), os_out))
                new_active_count = new_active_count + 1
            else:
                load_device = ""
                mod_device = ""
                if device == 57:
                    if read_chip_id(bus, device, 146) == 0xAB:
                        mod_device = "apds9960"
                    else:
                        mod_device = "veml6070"

                elif device == 64:
                    #print("hello64")
                    chip_id = read_chip_id(bus, device, 255)
                    #print("chipid = {0}".format(chip_id))
                    if chip_id == 0x1000 or chip_id == 0x1050:
                        mod_device = "hdc100x"
                    else:
                        mod_device = "htu21"

                elif ((device == 118) and (64 not in new_active)) or (device == 119):
                   chip_id = read_chip_id(bus, device, 208)
                   if chip_id > 0:
                       load_device = bosch_chip_id[read_chip_id(bus, device, 208)]
                       if load_device == "bme680":
                           mod_device = "bme680-i2c"
                       else:
                           mod_device = "bmp280-i2c"

                if mod_device != "":
                    if load_device == "":
                        load_device = mod_device
                    subprocess.run(["modprobe", mod_device])
                    print("Loading device {0} (chip ID {1}) on address {2}.".format(mod_device, chip_id, hex(device)))
                    new_device = "echo {0} {1} > /sys/bus/i2c/devices/i2c-1/new_device".format(load_device, hex(device))
                    os_out = os.system(new_device)
                    if os_out > 0:
                        print("New device exit code: {0}".format(os_out))
                    new_active_count = new_active_count + 1

        print("Loaded {0} of {1} device(s) found".format(new_active_count, device_count))
        bus.close()
        bus = None
        return new_active_count

    else:  # no devices found
        bus.close()
        bus = None
        return 0
    
